File: 2020/15_RambunctiousRecitation/test_memory.py
# ======================================================================
# Rambunctious Recitation
#   Advent of Code 2020 Day 15 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                    t e s t _ m e m o r y . p y
# ======================================================================
"Test class for Advent of Code 2020 day 15, Rambunctious Recitation"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import memory

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
EXAMPLE_TEXT = "0,3,6"
EXAMPLE_AGES = "0,3,3,1,0,4,0"

# ======================================================================
#                                                             TestMemory
# ======================================================================


class TestMemory(unittest.TestCase):  # pylint: disable=R0904
    "Test Memory object"

    def test_empty_init(self):
        "Test the default Memory creation"

        # 1. Create default Memory object
        myobj = memory.Memory()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(myobj.text, None)
        self.assertEqual(myobj.age, None)
        self.assertEqual(myobj.turn, 0)
        self.assertEqual(myobj.numbers, {})

    def test_text_init(self):
        "Test the Memory object creation from text"

        # 1. Create Memory object from text
        myobj = memory.Memory(text=EXAMPLE_TEXT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(len(myobj.text), 5)
        self.assertEqual(myobj.turn, 3)
        self.assertEqual(len(myobj.numbers), 3)
        self.assertEqual(myobj.age, 0)

        # 3. Speak more numbers
        for age in [int(_) for _ in EXAMPLE_AGES.split(',')]:
            self.assertEqual(myobj.age, age)
            myobj.add(age)

        # 4. Final check
        self.assertEqual(myobj.turn, 10)
        self.assertEqual(len(myobj.numbers), 5)
        self.assertEqual(myobj.age, 2)

    def test_add_last(self):
        "Test the Memory object creation from text"

        # 1. Create Memory object from text
        myobj = memory.Memory(text=EXAMPLE_TEXT)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.part2, False)
        self.assertEqual(len(myobj.text), 5)
        self.assertEqual(myobj.turn, 3)
        self.assertEqual(len(myobj.numbers), 3)
        self.assertEqual(myobj.age, 0)

        # 3. Speak more numbers
        for age in [int(_) for _ in EXAMPLE_AGES.split(',')]:
            self.assertEqual(myobj.age, age)
            myobj.add_last_spoken()

        # 4. Final check
        self.assertEqual(myobj.turn, 10)
        self.assertEqual(len(myobj.numbers), 5)
        self.assertEqual(myobj.age, 2)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                    t e s t _ m e m o r y . p y                 end
# ======================================================================

File: 2020/15_RambunctiousRecitation/test_memory.lua
-- ======================================================================
-- Rambunctious Recitation
--   Advent of Code 2020 Day 15 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ m e m o r y . l u a
-- ======================================================================
-- Test Memory for Advent of Code 2020 day 15, Rambunctious Recitation

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local memory = require('memory')

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = "0,3,6"
local EXAMPLE_AGES = "0,3,3,1,0,4,0"

-- ======================================================================
--                                                             TestMemory
-- ======================================================================

function test_empty_init()
  -- Test the default Memory creation

  -- 1. Create default Memory object
  local myobj = memory:Memory()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(myobj.age, nil)
  luaunit.assertEquals(myobj.turn, 0)
  luaunit.assertEquals(#myobj.numbers, 0)

end

function test_text_init()
  -- Test the Memory object creation from text

  -- 1. Create Game object from text
  local myobj = memory:Memory({text=EXAMPLE_TEXT})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 5)
  luaunit.assertEquals(myobj.age, 0)
  luaunit.assertEquals(myobj.turn, 3)
  luaunit.assertEquals(myobj.numbers[0], 1)
  luaunit.assertEquals(myobj.numbers[3], 2)
  luaunit.assertEquals(myobj.numbers[6], 3)
  
  -- 3. Speak more numbers
  for age in EXAMPLE_AGES:gmatch("[0-9]+") do
    age = tonumber(age)
    luaunit.assertEquals(myobj.age, age)
    myobj:add(age)
  end
  
  -- 4. Final check
  luaunit.assertEquals(myobj.turn, 10)
  luaunit.assertEquals(myobj.age, 2)
  
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                   t e s t _ m e m o r y . p y                  end
-- ======================================================================
